File: packages/firebase_common/lib/src/firestore_documents/force_update_config.flutterfire_gen.dart
// coverage:ignore-file
// ignore_for_file: type=lint
// ignore_for_file: unused_element, deprecated_member_use, deprecated_member_use_from_same_package, use_function_type_syntax_for_parameters, unnecessary_const, avoid_init_to_null, invalid_override_different_default_values_named, prefer_expression_function_bodies, annotate_overrides, invalid_annotation_target, unnecessary_question_mark

part of 'force_update_config.dart';

class ReadForceUpdateConfig {
  const ReadForceUpdateConfig._({
    required this.forceUpdateConfigId,
    required this.forceUpdateConfigReference,
    required this.iOSLatestVersion,
    required this.iOSMinRequiredVersion,
    required this.iOSForceUpdate,
    required this.androidLatestVersion,
    required this.androidMinRequiredVersion,
    required this.androidForceUpdate,
  });

  final String forceUpdateConfigId;
  final DocumentReference<ReadForceUpdateConfig> forceUpdateConfigReference;
  final String iOSLatestVersion;
  final String iOSMinRequiredVersion;
  final bool iOSForceUpdate;
  final String androidLatestVersion;
  final String androidMinRequiredVersion;
  final bool androidForceUpdate;

  factory ReadForceUpdateConfig._fromJson(Map<String, dynamic> json) {
    return ReadForceUpdateConfig._(
      forceUpdateConfigId: json['forceUpdateConfigId'] as String,
      forceUpdateConfigReference: json['forceUpdateConfigReference']
          as DocumentReference<ReadForceUpdateConfig>,
      iOSLatestVersion: json['iOSLatestVersion'] as String,
      iOSMinRequiredVersion: json['iOSMinRequiredVersion'] as String,
      iOSForceUpdate: json['iOSForceUpdate'] as bool? ?? false,
      androidLatestVersion: json['androidLatestVersion'] as String,
      androidMinRequiredVersion: json['androidMinRequiredVersion'] as String,
      androidForceUpdate: json['androidForceUpdate'] as bool? ?? false,
    );
  }

  factory ReadForceUpdateConfig.fromDocumentSnapshot(DocumentSnapshot ds) {
    final data = ds.data()! as Map<String, dynamic>;
    return ReadForceUpdateConfig._fromJson(<String, dynamic>{
      ...data,
      'forceUpdateConfigId': ds.id,
      'forceUpdateConfigReference':
          ds.reference.parent.doc(ds.id).withConverter(
                fromFirestore: (ds, _) =>
                    ReadForceUpdateConfig.fromDocumentSnapshot(ds),
                toFirestore: (obj, _) => throw UnimplementedError(),
              ),
    });
  }

  ReadForceUpdateConfig copyWith({
    String? forceUpdateConfigId,
    DocumentReference<ReadForceUpdateConfig>? forceUpdateConfigReference,
    String? iOSLatestVersion,
    String? iOSMinRequiredVersion,
    bool? iOSForceUpdate,
    String? androidLatestVersion,
    String? androidMinRequiredVersion,
    bool? androidForceUpdate,
  }) {
    return ReadForceUpdateConfig._(
      forceUpdateConfigId: forceUpdateConfigId ?? this.forceUpdateConfigId,
      forceUpdateConfigReference:
          forceUpdateConfigReference ?? this.forceUpdateConfigReference,
      iOSLatestVersion: iOSLatestVersion ?? this.iOSLatestVersion,
      iOSMinRequiredVersion:
          iOSMinRequiredVersion ?? this.iOSMinRequiredVersion,
      iOSForceUpdate: iOSForceUpdate ?? this.iOSForceUpdate,
      androidLatestVersion: androidLatestVersion ?? this.androidLatestVersion,
      androidMinRequiredVersion:
          androidMinRequiredVersion ?? this.androidMinRequiredVersion,
      androidForceUpdate: androidForceUpdate ?? this.androidForceUpdate,
    );
  }
}

class CreateForceUpdateConfig {
  const CreateForceUpdateConfig({
    required this.iOSLatestVersion,
    required this.iOSMinRequiredVersion,
    this.iOSForceUpdate = false,
    required this.androidLatestVersion,
    required this.androidMinRequiredVersion,
    this.androidForceUpdate = false,
  });

  final String iOSLatestVersion;
  final String iOSMinRequiredVersion;
  final bool iOSForceUpdate;
  final String androidLatestVersion;
  final String androidMinRequiredVersion;
  final bool androidForceUpdate;

  Map<String, dynamic> toJson() {
    return {
      'iOSLatestVersion': iOSLatestVersion,
      'iOSMinRequiredVersion': iOSMinRequiredVersion,
      'iOSForceUpdate': iOSForceUpdate,
      'androidLatestVersion': androidLatestVersion,
      'androidMinRequiredVersion': androidMinRequiredVersion,
      'androidForceUpdate': androidForceUpdate,
    };
  }
}

class UpdateForceUpdateConfig {
  const UpdateForceUpdateConfig({
    this.iOSLatestVersion,
    this.iOSMinRequiredVersion,
    this.iOSForceUpdate,
    this.androidLatestVersion,
    this.androidMinRequiredVersion,
    this.androidForceUpdate,
  });

  final String? iOSLatestVersion;
  final String? iOSMinRequiredVersion;
  final bool? iOSForceUpdate;
  final String? androidLatestVersion;
  final String? androidMinRequiredVersion;
  final bool? androidForceUpdate;

  Map<String, dynamic> toJson() {
    return {
      if (iOSLatestVersion != null) 'iOSLatestVersion': iOSLatestVersion,
      if (iOSMinRequiredVersion != null)
        'iOSMinRequiredVersion': iOSMinRequiredVersion,
      if (iOSForceUpdate != null) 'iOSForceUpdate': iOSForceUpdate,
      if (androidLatestVersion != null)
        'androidLatestVersion': androidLatestVersion,
      if (androidMinRequiredVersion != null)
        'androidMinRequiredVersion': androidMinRequiredVersion,
      if (androidForceUpdate != null) 'androidForceUpdate': androidForceUpdate,
    };
  }
}

/// A [CollectionReference] to configurations collection to read.
final readForceUpdateConfigCollectionReference = FirebaseFirestore.instance
    .collection('configurations')
    .withConverter<ReadForceUpdateConfig>(
      fromFirestore: (ds, _) => ReadForceUpdateConfig.fromDocumentSnapshot(ds),
      toFirestore: (obj, _) => throw UnimplementedError(),
    );

/// A [DocumentReference] to forceUpdateConfig document to read.
DocumentReference<ReadForceUpdateConfig>
    readForceUpdateConfigDocumentReference({
  required String forceUpdateConfigId,
}) =>
        readForceUpdateConfigCollectionReference.doc(forceUpdateConfigId);

/// A [CollectionReference] to configurations collection to create.
final createForceUpdateConfigCollectionReference = FirebaseFirestore.instance
    .collection('configurations')
    .withConverter<CreateForceUpdateConfig>(
      fromFirestore: (ds, _) => throw UnimplementedError(),
      toFirestore: (obj, _) => obj.toJson(),
    );

/// A [DocumentReference] to forceUpdateConfig document to create.
DocumentReference<CreateForceUpdateConfig>
    createForceUpdateConfigDocumentReference({
  required String forceUpdateConfigId,
}) =>
        createForceUpdateConfigCollectionReference.doc(forceUpdateConfigId);

/// A [CollectionReference] to configurations collection to update.
final updateForceUpdateConfigCollectionReference = FirebaseFirestore.instance
    .collection('configurations')
    .withConverter<UpdateForceUpdateConfig>(
      fromFirestore: (ds, _) => throw UnimplementedError(),
      toFirestore: (obj, _) => obj.toJson(),
    );

/// A [DocumentReference] to forceUpdateConfig document to update.
DocumentReference<UpdateForceUpdateConfig>
    updateForceUpdateConfigDocumentReference({
  required String forceUpdateConfigId,
}) =>
        updateForceUpdateConfigCollectionReference.doc(forceUpdateConfigId);

/// A [CollectionReference] to configurations collection to delete.
final deleteForceUpdateConfigCollectionReference =
    FirebaseFirestore.instance.collection('configurations');

/// A [DocumentReference] to forceUpdateConfig document to delete.
DocumentReference<Object?> deleteForceUpdateConfigDocumentReference({
  required String forceUpdateConfigId,
}) =>
    deleteForceUpdateConfigCollectionReference.doc(forceUpdateConfigId);

/// A query manager to execute query against [ForceUpdateConfig].
class ForceUpdateConfigQuery {
  /// Fetches [ReadForceUpdateConfig] documents.
  Future<List<ReadForceUpdateConfig>> fetchDocuments({
    GetOptions? options,
    Query<ReadForceUpdateConfig>? Function(Query<ReadForceUpdateConfig> query)?
        queryBuilder,
    int Function(ReadForceUpdateConfig lhs, ReadForceUpdateConfig rhs)? compare,
  }) async {
    Query<ReadForceUpdateConfig> query =
        readForceUpdateConfigCollectionReference;
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    final qs = await query.get(options);
    final result = qs.docs.map((qds) => qds.data()).toList();
    if (compare != null) {
      result.sort(compare);
    }
    return result;
  }

  /// Subscribes [ForceUpdateConfig] documents.
  Stream<List<ReadForceUpdateConfig>> subscribeDocuments({
    Query<ReadForceUpdateConfig>? Function(Query<ReadForceUpdateConfig> query)?
        queryBuilder,
    int Function(ReadForceUpdateConfig lhs, ReadForceUpdateConfig rhs)? compare,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    Query<ReadForceUpdateConfig> query =
        readForceUpdateConfigCollectionReference;
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    var streamQs =
        query.snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamQs = streamQs.where((qs) => !qs.metadata.hasPendingWrites);
    }
    return streamQs.map((qs) {
      final result = qs.docs.map((qds) => qds.data()).toList();
      if (compare != null) {
        result.sort(compare);
      }
      return result;
    });
  }

  /// Fetches a specified [ReadForceUpdateConfig] document.
  Future<ReadForceUpdateConfig?> fetchDocument({
    required String forceUpdateConfigId,
    GetOptions? options,
  }) async {
    final ds = await readForceUpdateConfigDocumentReference(
      forceUpdateConfigId: forceUpdateConfigId,
    ).get(options);
    return ds.data();
  }

  /// Subscribes a specified [ForceUpdateConfig] document.
  Stream<ReadForceUpdateConfig?> subscribeDocument({
    required String forceUpdateConfigId,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    var streamDs = readForceUpdateConfigDocumentReference(
      forceUpdateConfigId: forceUpdateConfigId,
    ).snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamDs = streamDs.where((ds) => !ds.metadata.hasPendingWrites);
    }
    return streamDs.map((ds) => ds.data());
  }

  /// Adds a [ForceUpdateConfig] document.
  Future<DocumentReference<CreateForceUpdateConfig>> add({
    required CreateForceUpdateConfig createForceUpdateConfig,
  }) =>
      createForceUpdateConfigCollectionReference.add(createForceUpdateConfig);

  /// Sets a [ForceUpdateConfig] document.
  Future<void> set({
    required String forceUpdateConfigId,
    required CreateForceUpdateConfig createForceUpdateConfig,
    SetOptions? options,
  }) =>
      createForceUpdateConfigDocumentReference(
        forceUpdateConfigId: forceUpdateConfigId,
      ).set(createForceUpdateConfig, options);

  /// Updates a specified [ForceUpdateConfig] document.
  Future<void> update({
    required String forceUpdateConfigId,
    required UpdateForceUpdateConfig updateForceUpdateConfig,
  }) =>
      updateForceUpdateConfigDocumentReference(
        forceUpdateConfigId: forceUpdateConfigId,
      ).update(updateForceUpdateConfig.toJson());

  /// Deletes a specified [ForceUpdateConfig] document.
  Future<void> delete({
    required String forceUpdateConfigId,
  }) =>
      deleteForceUpdateConfigDocumentReference(
        forceUpdateConfigId: forceUpdateConfigId,
      ).delete();
}

File: functions/src/firestore-documents/forceUpdateConfig.flutterfireGen.ts
import * as admin from 'firebase-admin'
import {
    DocumentData,
    DocumentReference,
    DocumentSnapshot,
    Query,
    QueryDocumentSnapshot,
    QuerySnapshot,
    SetOptions,
    WriteResult
} from 'firebase-admin/firestore'

export class ReadForceUpdateConfig {
    constructor({
        forceUpdateConfigId,
        path,
        iOSLatestVersion,
        iOSMinRequiredVersion,
        iOSForceUpdate,
        androidLatestVersion,
        androidMinRequiredVersion,
        androidForceUpdate
    }: {
        forceUpdateConfigId: string
        path: string
        iOSLatestVersion: string
        iOSMinRequiredVersion: string
        iOSForceUpdate: boolean
        androidLatestVersion: string
        androidMinRequiredVersion: string
        androidForceUpdate: boolean
    }) {
        this.forceUpdateConfigId = forceUpdateConfigId
        this.path = path
        this.iOSLatestVersion = iOSLatestVersion
        this.iOSMinRequiredVersion = iOSMinRequiredVersion
        this.iOSForceUpdate = iOSForceUpdate
        this.androidLatestVersion = androidLatestVersion
        this.androidMinRequiredVersion = androidMinRequiredVersion
        this.androidForceUpdate = androidForceUpdate
    }

    readonly forceUpdateConfigId: string

    readonly path: string

    readonly iOSLatestVersion: string

    readonly iOSMinRequiredVersion: string

    readonly iOSForceUpdate: boolean

    readonly androidLatestVersion: string

    readonly androidMinRequiredVersion: string

    readonly androidForceUpdate: boolean

    private static fromJson(
        json: Record<string, unknown>
    ): ReadForceUpdateConfig {
        return new ReadForceUpdateConfig({
            forceUpdateConfigId: json[`forceUpdateConfigId`] as string,
            path: json[`path`] as string,
            iOSLatestVersion: json[`iOSLatestVersion`] as string,
            iOSMinRequiredVersion: json[`iOSMinRequiredVersion`] as string,
            iOSForceUpdate:
                (json[`iOSForceUpdate`] as boolean | undefined) ?? false,
            androidLatestVersion: json[`androidLatestVersion`] as string,
            androidMinRequiredVersion: json[
                `androidMinRequiredVersion`
            ] as string,
            androidForceUpdate:
                (json[`androidForceUpdate`] as boolean | undefined) ?? false
        })
    }

    static fromDocumentSnapshot(ds: DocumentSnapshot): ReadForceUpdateConfig {
        const data = ds.data()!
        const cleanedData: Record<string, unknown> = {}
        for (const [key, value] of Object.entries(data)) {
            cleanedData[key] = value === null ? undefined : value
        }
        return ReadForceUpdateConfig.fromJson({
            ...cleanedData,
            forceUpdateConfigId: ds.id,
            path: ds.ref.path
        })
    }
}

export class CreateForceUpdateConfig {
    constructor({
        iOSLatestVersion,
        iOSMinRequiredVersion,
        iOSForceUpdate,
        androidLatestVersion,
        androidMinRequiredVersion,
        androidForceUpdate
    }: {
        iOSLatestVersion: string
        iOSMinRequiredVersion: string
        iOSForceUpdate: boolean
        androidLatestVersion: string
        androidMinRequiredVersion: string
        androidForceUpdate: boolean
    }) {
        this.iOSLatestVersion = iOSLatestVersion
        this.iOSMinRequiredVersion = iOSMinRequiredVersion
        this.iOSForceUpdate = iOSForceUpdate
        this.androidLatestVersion = androidLatestVersion
        this.androidMinRequiredVersion = androidMinRequiredVersion
        this.androidForceUpdate = androidForceUpdate
    }

    readonly iOSLatestVersion: string

    readonly iOSMinRequiredVersion: string

    readonly iOSForceUpdate: boolean

    readonly androidLatestVersion: string

    readonly androidMinRequiredVersion: string

    readonly androidForceUpdate: boolean

    toJson(): Record<string, unknown> {
        return {
            iOSLatestVersion: this.iOSLatestVersion,
            iOSMinRequiredVersion: this.iOSMinRequiredVersion,
            iOSForceUpdate: this.iOSForceUpdate,
            androidLatestVersion: this.androidLatestVersion,
            androidMinRequiredVersion: this.androidMinRequiredVersion,
            androidForceUpdate: this.androidForceUpdate
        }
    }
}

export class UpdateForceUpdateConfig {
    constructor({
        iOSLatestVersion,
        iOSMinRequiredVersion,
        iOSForceUpdate,
        androidLatestVersion,
        androidMinRequiredVersion,
        androidForceUpdate
    }: {
        iOSLatestVersion?: string
        iOSMinRequiredVersion?: string
        iOSForceUpdate?: boolean
        androidLatestVersion?: string
        androidMinRequiredVersion?: string
        androidForceUpdate?: boolean
    }) {
        this.iOSLatestVersion = iOSLatestVersion
        this.iOSMinRequiredVersion = iOSMinRequiredVersion
        this.iOSForceUpdate = iOSForceUpdate
        this.androidLatestVersion = androidLatestVersion
        this.androidMinRequiredVersion = androidMinRequiredVersion
        this.androidForceUpdate = androidForceUpdate
    }

    readonly iOSLatestVersion?: string

    readonly iOSMinRequiredVersion?: string

    readonly iOSForceUpdate?: boolean

    readonly androidLatestVersion?: string

    readonly androidMinRequiredVersion?: string

    readonly androidForceUpdate?: boolean

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    toJson(): Record<string, any> {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const json: Record<string, any> = {}
        if (this.iOSLatestVersion != undefined) {
            json[`iOSLatestVersion`] = this.iOSLatestVersion
        }
        if (this.iOSMinRequiredVersion != undefined) {
            json[`iOSMinRequiredVersion`] = this.iOSMinRequiredVersion
        }
        if (this.iOSForceUpdate != undefined) {
            json[`iOSForceUpdate`] = this.iOSForceUpdate
        }
        if (this.androidLatestVersion != undefined) {
            json[`androidLatestVersion`] = this.androidLatestVersion
        }
        if (this.androidMinRequiredVersion != undefined) {
            json[`androidMinRequiredVersion`] = this.androidMinRequiredVersion
        }
        if (this.androidForceUpdate != undefined) {
            json[`androidForceUpdate`] = this.androidForceUpdate
        }

        return json
    }
}

export class DeleteForceUpdateConfig {}

/**
 * A Cloud Firestore object which ignores `undefined` properties.
 */
const db = admin.firestore()
db.settings({ ignoreUndefinedProperties: true })

/**
 * Provides a reference to the configurations collection for reading.
 */
export const readForceUpdateConfigCollectionReference = db
    .collection(`configurations`)
    .withConverter<ReadForceUpdateConfig>({
        fromFirestore: (ds: DocumentSnapshot): ReadForceUpdateConfig => {
            return ReadForceUpdateConfig.fromDocumentSnapshot(ds)
        },
        toFirestore: () => {
            throw new Error(
                `toFirestore is not implemented for ReadForceUpdateConfig`
            )
        }
    })

/**
 * Provides a reference to a forceUpdateConfig document for reading.
 * @param forceUpdateConfigId - The ID of the forceUpdateConfig document to read.
 */
export const readForceUpdateConfigDocumentReference = ({
    forceUpdateConfigId
}: {
    forceUpdateConfigId: string
}): DocumentReference<ReadForceUpdateConfig> =>
    readForceUpdateConfigCollectionReference.doc(forceUpdateConfigId)

/**
 * Provides a reference to the configurations collection for creating.
 */
export const createForceUpdateConfigCollectionReference = db
    .collection(`configurations`)
    .withConverter<CreateForceUpdateConfig>({
        fromFirestore: () => {
            throw new Error(
                `fromFirestore is not implemented for CreateForceUpdateConfig`
            )
        },
        toFirestore: (obj: CreateForceUpdateConfig): DocumentData => {
            return obj.toJson()
        }
    })

/**
 * Provides a reference to a forceUpdateConfig document for creating.
 * @param forceUpdateConfigId - The ID of the forceUpdateConfig document to read.
 */
export const createForceUpdateConfigDocumentReference = ({
    forceUpdateConfigId
}: {
    forceUpdateConfigId: string
}): DocumentReference<CreateForceUpdateConfig> =>
    createForceUpdateConfigCollectionReference.doc(forceUpdateConfigId)

/**
 * Provides a reference to the configurations collection for updating.
 */
export const updateForceUpdateConfigCollectionReference = db
    .collection(`configurations`)
    .withConverter<UpdateForceUpdateConfig>({
        fromFirestore: () => {
            throw new Error(
                `fromFirestore is not implemented for UpdateForceUpdateConfig`
            )
        },
        toFirestore: (obj: UpdateForceUpdateConfig): DocumentData => {
            return obj.toJson()
        }
    })

/**
 * Provides a reference to a forceUpdateConfig document for updating.
 * @param forceUpdateConfigId - The ID of the forceUpdateConfig document to read.
 */
export const updateForceUpdateConfigDocumentReference = ({
    forceUpdateConfigId
}: {
    forceUpdateConfigId: string
}): DocumentReference<UpdateForceUpdateConfig> =>
    updateForceUpdateConfigCollectionReference.doc(forceUpdateConfigId)

/**
 * Provides a reference to the configurations collection for deleting.
 */
export const deleteForceUpdateConfigCollectionReference = db
    .collection(`configurations`)
    .withConverter<DeleteForceUpdateConfig>({
        fromFirestore: () => {
            throw new Error(
                `fromFirestore is not implemented for DeleteForceUpdateConfig`
            )
        },
        toFirestore: (): DocumentData => {
            throw new Error(
                `toFirestore is not implemented for DeleteForceUpdateConfig`
            )
        }
    })

/**
 * Provides a reference to a forceUpdateConfig document for deleting.
 * @param forceUpdateConfigId - The ID of the forceUpdateConfig document to read.
 */
export const deleteForceUpdateConfigDocumentReference = ({
    forceUpdateConfigId
}: {
    forceUpdateConfigId: string
}): DocumentReference<DeleteForceUpdateConfig> =>
    deleteForceUpdateConfigCollectionReference.doc(forceUpdateConfigId)

/**
 * Manages queries against the configurations collection.
 */
export class ForceUpdateConfigQuery {
    /**
     * Fetches forceUpdateConfig documents.
     * @param queryBuilder - Function to modify the query.
     * @param compare - Function to sort the results.
     */
    async fetchDocuments({
        queryBuilder,
        compare
    }: {
        queryBuilder?: (
            query: Query<ReadForceUpdateConfig>
        ) => Query<ReadForceUpdateConfig>
        compare?: (
            lhs: ReadForceUpdateConfig,
            rhs: ReadForceUpdateConfig
        ) => number
    }): Promise<ReadForceUpdateConfig[]> {
        let query: Query<ReadForceUpdateConfig> =
            readForceUpdateConfigCollectionReference
        if (queryBuilder != undefined) {
            query = queryBuilder(query)
        }
        const qs: QuerySnapshot<ReadForceUpdateConfig> = await query.get()
        let result = qs.docs.map(
            (qds: QueryDocumentSnapshot<ReadForceUpdateConfig>) => qds.data()
        )
        if (compare != undefined) {
            result = result.sort(compare)
        }
        return result
    }

    /**
     * Fetches a specific forceUpdateConfig document.
     * @param forceUpdateConfigId - The ID of the forceUpdateConfig document to fetch.
     */
    async fetchDocument({
        forceUpdateConfigId
    }: {
        forceUpdateConfigId: string
    }): Promise<ReadForceUpdateConfig | undefined> {
        const ds = await readForceUpdateConfigDocumentReference({
            forceUpdateConfigId
        }).get()
        return ds.data()
    }

    /**
     * Adds a forceUpdateConfig document.
     * @param createForceUpdateConfig - The forceUpdateConfig details to add.
     */
    async add({
        createForceUpdateConfig
    }: {
        createForceUpdateConfig: CreateForceUpdateConfig
    }): Promise<DocumentReference<CreateForceUpdateConfig>> {
        return createForceUpdateConfigCollectionReference.add(
            createForceUpdateConfig
        )
    }

    /**
     * Sets a forceUpdateConfig document.
     * @param forceUpdateConfigId - The ID of the forceUpdateConfig document to set.
     * @param createForceUpdateConfig - The forceUpdateConfig details to set.
     * @param options - Options for the set operation.
     */
    async set({
        forceUpdateConfigId,
        createForceUpdateConfig,
        options
    }: {
        forceUpdateConfigId: string
        createForceUpdateConfig: CreateForceUpdateConfig
        options?: SetOptions
    }): Promise<WriteResult> {
        if (options == undefined) {
            return createForceUpdateConfigDocumentReference({
                forceUpdateConfigId
            }).set(createForceUpdateConfig)
        } else {
            return createForceUpdateConfigDocumentReference({
                forceUpdateConfigId
            }).set(createForceUpdateConfig, options)
        }
    }

    /**
     * Updates a specific forceUpdateConfig document.
     * @param forceUpdateConfigId - The ID of the forceUpdateConfig document to update.
     * @param updateForceUpdateConfig - The details for updating the forceUpdateConfig.
     */
    async update({
        forceUpdateConfigId,
        updateForceUpdateConfig
    }: {
        forceUpdateConfigId: string
        updateForceUpdateConfig: UpdateForceUpdateConfig
    }): Promise<WriteResult> {
        return updateForceUpdateConfigDocumentReference({
            forceUpdateConfigId
        }).update(updateForceUpdateConfig.toJson())
    }

    /**
     * Deletes a specific forceUpdateConfig document.
     * @param forceUpdateConfigId - The ID of the forceUpdateConfig document to delete.
     */
    async delete({
        forceUpdateConfigId
    }: {
        forceUpdateConfigId: string
    }): Promise<WriteResult> {
        return deleteForceUpdateConfigDocumentReference({
            forceUpdateConfigId
        }).delete()
    }
}
